### ISSUE TITLE
Importing using oneclick in my team account doesn't get the map rendered the first time

### ISSUE THREAD

**Issue by saleiva** [CartoDB/cartodb#3817]
But if I go to the dashboard and click on it, it gets rendered perfectly.

**Comment by javisantana** [comment]
rendered the preview?

**Comment by saleiva** [comment]
Nope, sorry, the page itself. It renders some UI elements but not the map on the map view.

**Comment by juanignaciosl** [comment]
Could reproduce from http://catalog.data.gov/dataset/waofm-april-1-population-by-state-county-and-city-1990-to-present-22a1f, for example.

I get redirected to https://juanignaciosl.cartodb.com/viz/737522e8-23d9-11e5-b13b-0e43f3deba5a/map
And several requests fail with 401. The current problem is the URL. Opening the map through the dashboard redirects to https://team.cartodb.com/u/juanignaciosl/viz/737522e8-23d9-11e5-b13b-0e43f3deba5a/map, which is right. The failing one is throwing 401 because of [warden authentication](https://github.com/CartoDB/cartodb/blob/master/app/controllers/application_controller.rb#L103).

I'd like to know whether we should try to support this kind of urls for organization users (I don't think so, I guess this won't not work because warden scope) or we should fix [oneclick url generation](https://github.com/CartoDB/oneclick/blob/86bd866a617754de9278b1335918e34a98f89e09/routes/index.js#L64). A third way might be redirect to the right URL. @Kartones, what do you think? Since you've been working with warden, urls and multiuser you must have a great view on this :)

**Comment by Kartones** [comment]
well, warden cannot know about if a user is from org or not until performs the auth, but if the source is username.cartodb.com that auth code is ok (`CartoDB.extract_subdomain(request)` will give the username).

Why is it throwing a 401? It should redirect to the team.blablabla first (if the user is from an org but url hasn't got the org as subdomain) and then render the map, but the scope will always be the username (`extract_subdomain` is smart and uses /u/xxxx if present).

What I fear is the issue is that if we have private tables enabled, new maps will be private by default, so maybe they get imported but as they are private and you haven't specified an API key at that specific warden auth point, it throws back. But again not sure why we get to API auth if we have an open session

Without the full picture is hard to see, though.

If the origin JS knew about organizations and built the url as ORG.cartodb.com/u/USER it should produce same results, check https://github.com/CartoDB/cartodb/blob/master/config/initializers/carto_db.rb#L59 and https://github.com/CartoDB/cartodb/blob/master/config/initializers/carto_db.rb#L161. Flow is "if I have /u/USER_DOMAIN use it, else grab subdomain"

**Comment by juanignaciosl** [comment]
Related to CartoDB/cartodb#3127

**Comment by juanignaciosl** [comment]
Ok, I'll summarize here what I've learnt so far after speaking to @Kartones , @viddo and @rochoa regarding this issue and CartoDB/cartodb#3127.  Both are related to an authentication problem and CORS, as you'll see. Maybe we could make an ad-hoc approach for this one, but if we want to support both styles of URLs (see next point) we might need to change some code.

### How to reproduce:

1. Open a tab (with development tools opened) with the map view of a visualization you own. This URL looks like: `https://team.cartodb.com/u/juanignaciosl/viz/7917b784-df90-11e4-98b9-0e43f3deba5a/map`.
2. Open a new tab (with development tools opened) and change that URL to be username-based. In my example, `https://juanignaciosl.cartodb.com/viz/7917b784-df90-11e4-98b9-0e43f3deba5a/map`.

You'll see that in the second one several requests to layers and overlays endpoints fail with 401 error. 

### Why this happens

You can now compare the failing ones with the working ones in the other tab. As you can see, both are done against `team.cartodb...`. The main difference is that working ones send the cookies, including session cookies, and that's why they work.

This is not a cookie domain issue: if you take a look at the session cookie its domain is subdomain-free: `.cartodb.com`.

### Solution approach

The problem is CORS-related: since we're under different subdomains, browser behaves slighly different. More specifically, "[Standard CORS requests do not send or set any cookies by default](http://www.html5rocks.com/en/tutorials/cors/)".

In order to make XmlHttpRequest send them, we can do `xhr.withCredentials = true;`. I've added it [at the override of Backbone.sync](https://github.com/CartoDB/cartodb/blob/master/lib/assets/javascripts/cartodb/app.js#L51) and now that requests work. I think this is a legit solution, since we always want to send session cookies for authentication to work with "username-based URLs", and it's not a security concern, since we're only sending cookies under our domain, to our servers.

But now another problem arises: the response of a request to the SQL API is blocked by the browser with this error:
```
XMLHttpRequest cannot load 
http://central-org-b-admin.localhost.lan:8080/api/v2/sql?rows_per_page=40&p…from%20duplicated_column_name)%20__wrapped%20order%20by%20cartodb_id%20asc. 
A wildcard '*' cannot be used in the 'Access-Control-Allow-Origin' header 
when the credentials flag is true. 
Origin 'http://central-org-b.localhost.lan:3000' is therefore not allowed access.
```

The problem is that with `Access-Control-Allow-Origin`, "["*" cannot be used for a resource that supports credentials.](http://www.w3.org/TR/cors/)". By adding `xhr.withCredentials = true;` we've fixed authentication issues, but responses from servers with `Access-Control-Allow-Origin:*` won't work.

So, after speaking to @rochoa , we think this should be solved at frontend this way: before making a request to the server endpoint, request OPTIONS in order to check whether it has `Access-Control-Allow-Origin:*` or not, and set `xhr.withCredentials = true;` in requests made to servers without that configuration.

@rambo, @viddo, what do you think? I is right and feasible?

**Comment by javisantana** [comment]
as I see this:

- it sounds like a lot of things to change because a bug in oneclick
- I think it should be possible to redirect from user.cartodb.com to org.cartodb.com/u/user. If you take a look at the page loaded with user.cartodb.com the base_url is the organization one so backend **knows** it's a organization user. I don't get the @Kartones related to the redirection, could you explain

**Comment by javisantana** [comment]
more data:

if ``base_url`` url is correct everything works as expected, I just tried it changing it debugging the app, so:

If you check it for user.cartodb.com the base_url use org.cartodb.com/u/user and should be user.cartodb.com.

**Comment by juanignaciosl** [comment]
- This issue is onclick bug (I've already talked about an ad-hoc solution, probably on its side), but the problem underneath is bigger and already reported at CartoDB/cartodb#3127.
- In most cases that redirection works, performed [in this line](https://github.com/CartoDB/cartodb/blob/master/app/controllers/application_controller.rb#L191), but it's not because [that endpoint has `dont_rewrite=true`](https://github.com/CartoDB/cartodb/blob/fc4798f61c38be19ede3c93d8837ba07fa309ef8/config/routes.rb#L254), in order for old links not to be broken [(explanation)](https://github.com/CartoDB/cartodb/blob/fc4798f61c38be19ede3c93d8837ba07fa309ef8/config/routes.rb#L6). We could modify the check for dont_rewrite parameter and add `dont_rewrite=false` in oneclick if we prefer an specific solution.

**Comment by javisantana** [comment]
The base_url thing is the solution for #3217, right? instead of changing headers and dealing with CORS, OPTIONS, cookie domains and javascript stuff, why don't just set the correct url?

FYI (spanish) http://javisantana.tumblr.com/post/61488510961/cors-en-la-vida-real

**Comment by juanignaciosl** [comment]
Explaining what @javisantana said for those who didn't know, like me:

Non-working urls have `user_data.base_url` pointing to organization URL, and CORS-related issues begin there. His proposal is making it point to user url if that's the one user requested (avoiding all other issues then).

**Comment by Kartones** [comment]
Really careful if you change that, test deeply having shared tables, etc. at organizations, that they still work  at the dashboard etc., because IIRC that was one of the points of having base_url (I'd have to check but probably there are more), public pages (Maps list, etc. and maybe also the "datasets used in this visualization" of come from multiple users).

The real user url for an org user is `ORG.domain/u/USERNAME`, that's why it comes at `base_url` that way. Changing it for oneclick for me is a bit flipping over the issue to the other side. If there's scenarios where base_url does not "work" for the intended use, why not split it into two alternatives: `base_url` (current one) and `legacy_url` or `organizationless-url`?

**Comment by javisantana** [comment]
there are two issues:

- onclick not redirecting to the right url
- cartodb not knowing how to deal with both url types for an user (this one)

adding a new variable means changing the frontend to fix a problem is not frontend related. If we don't want to support user.cartodb.com for org users let's redirect.

What are the problems related with redirecting?

**Comment by Kartones** [comment]
The only problem I recall is not wanting to redirect on old visualizations (https://github.com/CartoDB/cartodb/issues/3127#issuecomment-91601539), that **regarding the `dont_rewrite` MVC argument**. Again I just implemented that but don't know specifics (and we didn't documented them), so fine to do it, if we break old vis from andrew or saleiva we'll notice and revert.

Changing `user_data.base_url` might trigger issues on organization accounts where the FE builds urls (dashboard with shared stuff, org public pages, etc.). If it works, fine by me, Is set that parameter for the frontend so it knows how to build urls. I'm just warning that changing it is not just "redirecting", is probably affecting Backbone urls generation... (@viddo, am I right?).

**Comment by juanignaciosl** [comment]
@Kartones do you know how will we notice (other than their complains) if we remove redirecting?

@javisantana , @Kartones , do you agree on allowing redirection only for the two reported problematic urls, `get '(/user/:user_domain)(/u/:user_domain)/tables/:id/` and `get '(/user/:user_domain)(/u/:user_domain)/viz/:id/map`?

**Comment by Kartones** [comment]
no idea, maybe an increase of 30x at our stats... but the app itself should behave as expected (just try with other org user urls,etc. to be sure but...).

For me ok.
